lib/using_restricted_array.rb

```diff
@@ -6,58 +6,104 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(1)
 def length(array)
-  raise NotImplementedError
+  length = 0
+  i = 0
+  while array[i] != nil
+    length += 1
+    i+=1
+  end 
+  return length
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: n/a
 def print_array(array)
-  raise NotImplementedError
+  array.each do |i|
+    puts i 
+  end 
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  length.times do |i|
+    if array[i] == value_to_find 
+      return true
+    end
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  largest = array[0]
+  length.times do |i|
+    if largest <= array[i]
+      largest = array[i]
+    end 
+  end 
+  return largest
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  smallest = array[0]
+
+  length.times do |i|
+    if smallest >= array[i]
+      smallest = array[i]
+    end 
+  end 
+  return smallest
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n/2)
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  (length/2).times do |index|
+    temp = array[index]
+    array[index] = array[length - 1 - index]
+    array[length - 1 - index] = temp
+  end 
+  return array 
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(1) or o(n)
+# Space complexity: o(1)
 def binary_search(array, length, value_to_find)
```

Comment:
👍 This works, but the time complexity is O(log n) because we cut the array in half with each iteration.